### ISSUE TITLE
no-undef rule applies to Flow type annotations

### ISSUE THREAD

**Issue by tmcw** [eslint/eslint#4741]
In my codebase I have the lines:

```js
  warnInProgressUpload(e: UnloadEvent) {
    if (!this.state.uploadingUploads.size) return undefined;
    var confirmationMessage = 'You have an upload in progress, if you leave the page your upload will be cancelled.';
    (e || window.event).returnValue = confirmationMessage;
    return confirmationMessage;
  }
```

`UnloadEvent` is a [type declaration defined in a separate file](http://flowtype.org/docs/declarations.html#_)

```js
declare class UnloadEvent extends Event {
    returnValue: string;
}
```

This produces the error:

```
  107:27  error  "UnloadEvent" is not defined  no-undef
```

Can eslint skip running no-undef on type annotations?

**Comment by eslintbot** [comment]
@tmcw Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by gyandeeps** [comment]
We dont support type annotations out of the box in eslint. With that being said if I run eslint your code then it will give me a parser error.
Can you plz give us more details on what parser are you using and any other relevant information?

**Comment by ilyavolodin** [comment]
ESLint doesn't support flow at all, so it has no knowledge of type annotations. You might want to create an issue for [`eslint-plugin-babel`](https://github.com/babel/eslint-plugin-babel)

@gyandeeps `babel-eslint` supports flow, so I assume it was used as a parser.

**Comment by tmcw** [comment]
Ah yes, sorry:

```
~/src/app〉./node_modules/.bin/eslint -v
v1.10.3
```

And using babel-eslint 5.0.0-beta6

**Comment by hzoo** [comment]
@tmcw https://github.com/babel/babel-eslint#known-issues

**Comment by tmcw** [comment]
Ah thanks everyone!

**Comment by gibbok** [comment]
Check out eslint-plugin-flowtype: An eslint plugin that makes flow type annotations global variables and marks declarations as used. Solves the problem of false positives with no-undef and no-unused-vars.
